### PR TITLE
v1.4 Parse Date and Time separately for Reservations

### LINKED ISSUE #205: Intuitiveness of reservation command
**Describe the bug**
To add a reservation, we must provide the timestamp in the format of "2018-12-03T10:00:00" which is not user intuitive.

**Expected behavior**
Ideally, the command should parse natural language input, or make use of two prefixes:

`add-reservation n/John Doe px/4 dt/12th october 2018 12pm t/Driving`
Can be achieved using Natty library.
http://natty.joestelmach.com/doc.jsp

or

`add-reservation n/John Doe px/4 d/12-10-18 t/23:59 t/Driving`

### LINKED ISSUE #207: Reservation timestamp should be validated
**Describe the bug**
Creating reservation can allow us to input timestamp that has already passed. 

**To Reproduce**
1. Look at current timestamp
2. Attempt to add a reservation whose timestamp has already passed, and it gets successfully added.

**Expected behavior**
Do not allow reservation to be added if timestamp has already passed current timestamp.

We assume using system time.

### PR DESCRIPTION
**Pull request by m4dkip**
As mentioned in #205, this PR splits the `DateTime` parameter into `Date` and `Time` for Reservation related commands.

~~This PR is still WIP, I am still planning to:~~

~~1. Update tests~~
~~2. Implement NATTY to parse the date and time. (#205)~~
~~3. Add a check to current system date to prevent adding reservations on dates that have already passed (#207).~~

PR finished! Ready for review.

Closes #205 
Closes #207

### CODE REVIEW

**Comment by m4dkip** [comment]
I left the Tag functionality there in case it would break the dependencies.

Maybe the owner could use the Tag feature to assign Table numbers? 

Or maybe tag a bunch of multiple reservations under "Corporate Event". Then he's able to quickly see who's under the corporate event at a glance.

**Comment by AZhiKai** [comment]
That would be a better use than using tags that don't make sense. Do keep  that in mind as we need to populate the data before our 2103 demo (or in fact even for 2101 since we using screenshots). :smile:

File: src/main/java/seedu/restaurant/model/reservation/Date.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+     *
+     * !!!!!!!!!!!! IMPORTANT NOTE !!!!!!!!!!!!!
+     * Firstly checks if the date is able to be regularly parsed in DD-MM-YYYY format without using NATTY. This is
+     * necessary as NATTY cannot be configured to interpret dates in DD-MM-YYYY format, and will default to MM-DD-YYYY.
+     *
+     * i.e. NATTY will think 03-12-2018 means "12th of March" instead of "3rd of December"
+     *
+     * @param date A valid date.
+     */
+    public Date(String date) {
```
Comment (m4dkip) [inline src/main/java/seedu/restaurant/model/reservation/Date.java]: I think I implemented this method quite poorly, any ideas on how to make this neater?

Comment (AZhiKai) [reply src/main/java/seedu/restaurant/model/reservation/Date.java]: Hmm this is intriguing. I would have thought Natty would be able to set the the format. Is there any documentation on how to change the format? Try Stackoverflow?

Comment (m4dkip) [reply src/main/java/seedu/restaurant/model/reservation/Date.java]: Have been trying to search for a solution for quite a while now, can't seem to find anything.

Comment (AZhiKai) [reply src/main/java/seedu/restaurant/model/reservation/Date.java]: Hm I'm not quite sure as I didn't use the library before. Will try sniffing around the goods once I'm done with the Person stuffs.

Comment (m4dkip) [reply src/main/java/seedu/restaurant/model/reservation/Date.java]: Thank you!!

**Comment by AZhiKai** [comment]
@m4dkip  Do add tests for the removeTag methods you added. :)

File: docs/UserGuide.adoc [hunk 2]
```diff
@@ -626,16 +626,28 @@ Add/Update the recipe of the 1st item in the results of the `find` command.
...
 Adds a new reservation to the reservations list. +
-Format: `add-reservation n/NAME px/PAX dt/DATETIME` or `ar n/NAME px/PAX dt/DATETIME`
+Format: `add-reservation n/NAME px/PAX d/DATE ti/TIME` or `ar n/NAME px/PAX d/DATE ti/TIME`
 
-****
-* DateTime is entered in yyyy-mm-dd**T**hh:mm:ss format.
-****
+[NOTE]
+====
+* We use Natural Language processing to parse the date and time values.
```
Comment (AZhiKai) [inline docs/UserGuide.adoc]: I feel "Natural Language" should just be "natural language processing". :P

Comment (m4dkip) [reply docs/UserGuide.adoc]: I agree! Made the change.

File: docs/UserGuide.adoc [hunk 3]
```diff
@@ -626,16 +626,28 @@ Add/Update the recipe of the 1st item in the results of the `find` command.
 ==== Adding a reservation: `add-reservation` or `ar`
 
 Adds a new reservation to the reservations list. +
-Format: `add-reservation n/NAME px/PAX dt/DATETIME` or `ar n/NAME px/PAX dt/DATETIME`
+Format: `add-reservation n/NAME px/PAX d/DATE ti/TIME` or `ar n/NAME px/PAX d/DATE ti/TIME`
```
Comment (AZhiKai) [inline docs/UserGuide.adoc]: is prefix "t/" taken?

File: src/test/data/XmlSerializableRestaurantBookTest/duplicateReservationOnlyRestaurantBook.xml [hunk 4]
```diff
@@ -3,15 +3,17 @@
     <reservations>
         <name>Jeeves</name>
         <pax isPrivate="false">4</pax>
-        <dateTime isPrivate="false">2018-11-03T10:15:00</dateTime>
+        <date isPrivate="false">03-11-2019</date>
```
Comment (AZhiKai) [inline src/test/data/XmlSerializableRestaurantBookTest/duplicateReservationOnlyRestaurantBook.xml]: Why is there a `isPrivate` field here?

File: src/test/data/XmlSerializableRestaurantBookTest/duplicateReservationOnlyRestaurantBook.xml [hunk 5]
```diff
@@ -3,15 +3,17 @@
...
+        <time isPrivate="false">10:15</time>
         <remark isPrivate="false">Some remarks</remark>
     </reservations>
 
     <!-- Person with same identity as Jeeves -->
```
Comment (AZhiKai) [inline src/test/data/XmlSerializableRestaurantBookTest/duplicateReservationOnlyRestaurantBook.xml]: Correct comment

File: docs/UserGuide.adoc (hunk 3 above)
Comment (m4dkip) [reply docs/UserGuide.adoc]: Yes, t/ is already taken up by `Tag`

File: src/test/data/XmlSerializableRestaurantBookTest/duplicateReservationOnlyRestaurantBook.xml (hunk 4 above)
Comment (m4dkip) [reply src/test/data/XmlSerializableRestaurantBookTest/duplicateReservationOnlyRestaurantBook.xml]: Good catch, will remove it

File: docs/UserGuide.adoc (hunk 3 above)
Comment (AZhiKai) [reply docs/UserGuide.adoc]: Okay!

File: src/test/java/seedu/restaurant/model/reservation/DateTest.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        Assert.assertThrows(IllegalArgumentException.class, () -> new Date(invalidDate));
+    }
+
+    @Test
+    public void isValidDate() {
+        // null date
+        Assert.assertThrows(NullPointerException.class, () -> Date.isValidDate(null));
+
+        // invalid dates
+        assertFalse(Date.isValidDate("")); // empty string
```
Comment (HyperionNKJ) [inline src/test/java/seedu/restaurant/model/reservation/DateTest.java]: I thought it'd be good to include a date that does not exist. Like "31-02-2018"

File: src/test/java/seedu/restaurant/logic/commands/CommandTestUtil.java [hunk 7]
```diff
@@ -214,16 +218,26 @@
...
+            " " + PREFIX_TIME + VALID_RESERVATION_TIME_BILLY;
     public static final String RESERVATION_TAG_DESC_ANDREW = " " + PREFIX_TAG + VALID_RESERVATION_TAG_ANDREW;
     public static final String RESERVATION_TAG_DESC_BILLY = " " + PREFIX_TAG + VALID_RESERVATION_TAG_BILLY;
 
     public static final String INVALID_RESERVATION_NAME_DESC = " " + PREFIX_NAME + "S&shrew"; // '&' not allowed
     public static final String INVALID_RESERVATION_PAX_DESC = " " + PREFIX_PAX + "a4"; // letters not allowed
-    public static final String INVALID_RESERVATION_DATETIME_DESC = " " + PREFIX_DATETIME + "2018-99"; // incomplete
+    //public static final String INVALID_RESERVATION_DATETIME_DESC = " " + PREFIX_DATETIME + "2018-99"; // incomplete
+    public static final String INVALID_RESERVATION_DATE_DESC = " " + PREFIX_DATE + "hi im a date"; // terrible joke
+    public static final String INVALID_RESERVATION_TIME_DESC = " " + PREFIX_TIME + "hi im a time"; // bad joke
```
[the remark this answers is not in the document]
Comment (AZhiKai) [reply src/test/java/seedu/restaurant/logic/commands/CommandTestUtil.java]: Is this intended? LOL

File: src/test/java/seedu/restaurant/model/reservation/DateTest.java (hunk 6 above)
Comment (AZhiKai) [reply src/test/java/seedu/restaurant/model/reservation/DateTest.java]: Or testing out leap year, as well.

File: docs/UserGuide.adoc [hunk 8]
```diff
@@ -627,16 +627,28 @@ Add/Update the recipe of the 1st item in the results of the `find` command.
...
-****
+[NOTE]
+====
+* We use natural language processing to parse the date and time values.
+* For example, phrases like `21st Dec` or `12 p.m.` will be accepted. Try it out!
+====
+
+[WARNING]
+====
+* Try to avoid ambiguous language as the it might be interpreted incorrectly.
```
[the remark this answers is not in the document]
Comment (m4dkip) [reply docs/UserGuide.adoc]: Good catch. Thanks!

File: src/main/java/seedu/restaurant/model/reservation/Date.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+//@@author m4dkip
+/**
+ * Represents a date in the calendar in DD-MM-YYYY format Guarantees: immutable; is valid as declared in {@link
+ * #isValidDate(String)}
+ */
+public class Date {
+
+    public static final String MESSAGE_DATE_CONSTRAINTS =
+            "Dates should only contain numbers, and it should be in the format DD-MM-YYYY.\nThe date must exist in "
```
[the remark this answers is not in the document]
Comment (m4dkip) [reply src/main/java/seedu/restaurant/model/reservation/Date.java]: Good idea.

File: src/test/java/seedu/restaurant/logic/commands/CommandTestUtil.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (m4dkip) [reply src/test/java/seedu/restaurant/logic/commands/CommandTestUtil.java]: LOL just wanted to make sure NATTY couldn't parse it, so I had to type some garbage

File: src/test/java/seedu/restaurant/logic/parser/reservation/AddReservationCommandParserTest.java [hunk 10]
```diff
@@ -84,48 +97,58 @@ public void parse_compulsoryFieldMissing_failure() {
...
+                        + VALID_RESERVATION_DATE_ANDREW + RESERVATION_TIME_DESC_ANDREW,
+                expectedMessage);
+
+        // missing time prefix
+        assertParseFailure(parser, RESERVATION_NAME_DESC_ANDREW + RESERVATION_PAX_DESC_ANDREW
+                        + RESERVATION_DATE_DESC_ANDREW + VALID_RESERVATION_TIME_ANDREW,
                 expectedMessage);
 
         // all prefixes missing
         assertParseFailure(parser, VALID_RESERVATION_NAME_ANDREW + VALID_RESERVATION_NAME_ANDREW
```
[the remark this answers is not in the document]
Comment (m4dkip) [reply src/test/java/seedu/restaurant/logic/parser/reservation/AddReservationCommandParserTest.java]: tHANK YOU

File: src/test/java/seedu/restaurant/model/reservation/DateTest.java (hunk 6 above)
Comment (m4dkip) [reply src/test/java/seedu/restaurant/model/reservation/DateTest.java]: Unfortunately, Natty accepts 31-02-2019 and tries to guess...

File: src/test/java/seedu/restaurant/logic/commands/CommandTestUtil.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (AZhiKai) [reply src/test/java/seedu/restaurant/logic/commands/CommandTestUtil.java]: Oh sounds good, didn't know it's for test. :P

File: src/test/java/seedu/restaurant/model/reservation/DateTest.java (hunk 6 above)
Comment (AZhiKai) [reply src/test/java/seedu/restaurant/model/reservation/DateTest.java]: I'm not surprised. I don't think they're able to handle the special dates.

**Pull request merged**